iis-wf/iis-wf-referenceextraction/src/main/java/eu/dnetlib/iis/wf/referenceextraction/patent/PatentMetadataRetrieverJob.java

```diff
@@ -52,50 +63,101 @@
+                JavaRDD<Fault> faults = returnedFromEPO.filter(e -> e._2.getException() != null).map(e -> FaultUtils.exceptionToFault(e._1, e._2.getException(), null));
+                
+                JavaRDD<DocumentText> entitiesToBeWritten;
                 
-                JavaRDD<DocumentText> retrievedPatentMeta = obtainedIdToResponse.map(e -> DocumentText.newBuilder().setId(e._1).setText(e._2.getContent()).build());
-                JavaRDD<Fault> faults = obtainedIdToResponse.filter(e -> e._2.getException() != null).map(e -> FaultUtils.exceptionToFault(e._1, e._2.getException(), null));
-                JavaRDD<ReportEntry> reportEntries = generateReportEntries(sc, retrievedPatentMeta, faults);
-                        
-                storeInOutput(retrievedPatentMeta, faults, reportEntries, outputPaths, params.numberOfEmittedFiles);
+                if (!returnedFromEPO.isEmpty()) {
+                    // storing new cache entry
+                    JavaRDD<Fault> cachedFaults = CacheStorageUtils.getRddOrEmpty(sc, avroLoader, cacheRootDir,
+                            existingCacheId, CacheRecordType.fault, Fault.class);
+
+                    CacheStorageUtils.storeInCache(avroSaver, cachedSources.union(retrievedPatentMeta),
+                            cachedFaults.union(faults), cacheRootDir, lockManager, cacheManager, hadoopConf,
+                            params.numberOfEmittedFiles);
+                    
+                    // merging final results
+                    entitiesToBeWritten = entitiesReturnedFromCache.union(retrievedPatentMeta);
+                    
+                } else {
+                    entitiesToBeWritten = entitiesReturnedFromCache;
+                }
                 
+                // store final results
+                storeInOutput(entitiesToBeWritten, 
+                        //notice: we do not propagate faults from cache, only new faults are written
+                        faults, generateReportEntries(sc, entitiesReturnedFromCache, retrievedPatentMeta, faults), 
+                        new OutputPaths(params), params.numberOfEmittedFiles);
+
+
             } catch (ServiceFacadeException e) {
                 throw new RuntimeException("unable to instantiate patent service facade!", e);
             }
         }
     }
 
     //------------------------ PRIVATE --------------------------
+        
+    private static JavaPairRDD<CharSequence, ContentRetrieverResponse> retriveFromRemoteEndpoint(JavaRDD<ImportedPatent> importedPatent,
+            PatentServiceFacade patentServiceFacade) {
+        return importedPatent
+                // limiting number of partitions to 1 in order to run EPO retrieval within a single task
+                .coalesce(1).mapToPair(x -> new Tuple2<>(getId(x), getMetadataFromFacade(x, patentServiceFacade)));
```

Comment:
Yes, same is limiting it on spark parameters level which was an alternative (https://github.com/openaire/iis/issues/1096#issuecomment-668113876). The problem is we want both: limiting the parallelism for EPO communication phase and parallel processing for all the other stages. 

The `coalesce(1)` here is the closest to the EPO retrieval operation it can be although it is still up to the execution plan to decide how far this 1 partition will affect the upstream processing (as already indicated in https://github.com/openaire/iis/issues/1103#issuecomment-656716431). Not sure how to enforce the plan to apply this transformation only to a single mapping step. Maybe if the higher parallelization level of the upstream operations will be justified the number of partitions will not be infected in those phases?

Reply:
OK, I have found an ultimate solution to enforce the plan to apply the repartition only to a single mapping step: `repartition(1)`. More details in edited comment: https://github.com/openaire/iis/issues/1096#issuecomment-668113876.

Reply:
Btw, I have also created to apply similar approach in `CachedWebCrawlerJob`: https://github.com/openaire/iis/issues/1131.

I don't want this PR to include vaguely related changes.

Reply:
That's the best solution for us right now :) However, I've been thinking of a more general solution, similar maybe to something like this [Rest Data Source for Apache Spark](https://github.com/sourav-mazumder/Data-Science-Extensions/tree/master/spark-datasource-rest). Another solution could be a distributed queue that would be fed from webservice endpoint and read by multiple spark executors. Any solution would be reasonable only if we could reuse it in many places because the amount of work for implementing the changes would probably be significant.

Reply:
Yeap, I agree, let's create dedicated ticket for this task. At least to explore the potential usage because as we have discussed during the call some use cases might benefit from it.